### PR TITLE
Add audit fields to Reports

### PR DESCRIPTION
**Pull request by ghickman**
This adds auditing fields to the Reports model, tracking who and when reports are added and updated.

The Django Admin has a log of events (using Generic FKs) which I've populated the records with.  I've updated the admin to both make use of this information for display and filtering, and to populate it on save.

### CODE REVIEW

File: reports/admin.py [hunk 1]
```diff
@@ -140,4 +140,13 @@ def doi_suffix(self, obj):  # pragma: no cover
         else:
             return "Save to generate DOI suffix"
 
+    def save_model(self, request, obj, form, change):
+        if obj.created_by_id is None:
+            obj.created_by = request.user
+
+        # ensure updated_at is both set on model creation _and_ update
```
Comment (lucyb) [inline reports/admin.py:147]: Small this, but the comment says `updated_at`, but the code refers to `updated_by`

File: reports/models.py [hunk 2]
```diff
@@ -189,6 +190,23 @@ class Report(models.Model):
...
+        null=True,
+        blank=True,
+    )
+
+    updated_at = models.DateTimeField(null=True)
+    updated_by = models.ForeignKey(
+        "gateway.User",
+        on_delete=models.PROTECT,
+        related_name="reports_updated",
+        null=True,
```
Comment (lucyb) [inline reports/models.py]: Should this also have `blank=True`?

File: reports/models.py [hunk 3]
```diff
@@ -190,21 +190,18 @@ class Report(models.Model):
         help_text="A description of why this report, from an external organisation, is on the OpenSAFELY Reports website.",
     )
 
-    created_at = models.DateTimeField(default=timezone.now, null=True)
+    created_at = models.DateTimeField(default=timezone.now)
```
Comment (lucyb) [inline reports/models.py:193]: Could `auto_now_add` be used here or do you want to preserve the ability to modify it?

**Review by lucyb — CHANGES_REQUESTED** [review]
This is probably something about my local setup, but when I try to apply the migration locally, I get the following error. Do you know why?

```
(.venv) ➜  reports.opensafely.org git:(add-audit-fields) ✗ python manage.py migrate         
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, gateway, reports, sessions
Running migrations:
  Applying reports.0031_get_audit_details_from_admin_log_entries...Traceback (most recent call last):
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/manage.py", line 22, in <module>
    main()
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/base.py", line 402, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/base.py", line 448, in execute
    output = self.handle(*args, **options)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/base.py", line 96, in wrapped
    res = handle_func(*args, **kwargs)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/commands/migrate.py", line 349, in handle
    post_migrate_state = executor.migrate(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/migration.py", line 130, in apply
    operation.database_forwards(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/reports/migrations/0031_get_audit_details_from_admin_log_entries.py", line 24, in copy_audit_data_from_log_entries_to_reports
    report.created_at = log.action_time
AttributeError: 'NoneType' object has no attribute 'action_time'
```

File: reports/models.py (hunk 3 above)
Comment (ghickman) [reply reports/models.py:193]: Never use `auto_now_add`!  It removes the ability to change the value, while default can be overridden.

File: reports/models.py (hunk 2 above)
Comment (ghickman) [reply reports/models.py]: No, this is a read-only field in the admin so we don't need to care about validating it there.

**Comment by ghickman** [comment]
@lucyb – I'm guessing you didn't apply that migration on top of a prod db dump?  I'll make that part of the migration optional becaus we still want them to work with an empty database.

**Comment by lucyb** [comment]
> I'm guessing you didn't apply that migration on top of a prod db dump? I'll make that part of the migration optional because we still want them to work with an empty database.

I ran the `populate_reports` command rather than take a prod db dump (it's the default behaviour for the dev docker container).

Thank you!

**Comment by lucyb** [comment]
That fixed one migration. This is the next error:

```
(.venv) ➜  reports.opensafely.org git:(add-audit-fields) ✗ python manage.py migrate
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, gateway, reports, sessions
Running migrations:
  Applying reports.0031_get_audit_details_from_admin_log_entries... OK
  Applying reports.0032_require_created_and_updated_fields...Traceback (most recent call last):
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/sqlite3/base.py", line 357, in execute
    return Database.Cursor.execute(self, query, params)
sqlite3.IntegrityError: NOT NULL constraint failed: new__reports_report.created_by_id

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/manage.py", line 22, in <module>
    main()
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/base.py", line 402, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/base.py", line 448, in execute
    output = self.handle(*args, **options)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/base.py", line 96, in wrapped
    res = handle_func(*args, **kwargs)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/core/management/commands/migrate.py", line 349, in handle
    post_migrate_state = executor.migrate(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/migration.py", line 130, in apply
    operation.database_forwards(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/migrations/operations/fields.py", line 235, in database_forwards
    schema_editor.alter_field(from_model, from_field, to_field)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/sqlite3/schema.py", line 174, in alter_field
    super().alter_field(model, old_field, new_field, strict=strict)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/base/schema.py", line 788, in alter_field
    self._alter_field(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/sqlite3/schema.py", line 457, in _alter_field
    self._remake_table(model, alter_field=(old_field, new_field))
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/sqlite3/schema.py", line 333, in _remake_table
    self.execute(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/base/schema.py", line 199, in execute
    cursor.execute(sql, params)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 103, in execute
    return super().execute(sql, params)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute
    with self.db.wrap_database_errors:
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
  File "/home/lucy/datalab/opensafely/reports.opensafely.org/.venv/lib/python3.10/site-packages/django/db/backends/sqlite3/base.py", line 357, in execute
    return Database.Cursor.execute(self, query, params)
django.db.utils.IntegrityError: NOT NULL constraint failed: new__reports_report.created_by_id
```

**Comment by lucyb** [comment]
As an aside, thank you for splitting up the changes into nice, discrete commits again. It was good to read through.

**Comment by ghickman** [comment]
Ok, this is a tricky one.  I can't recreate this with a copy of prod, an empty db, or a db with groups, a user, and reports populated.  I think this is your local db having Reports without matching LogEntry records and I'm not sure we should handle that since prod doesn't suffer from it and it blocks turn off nullable for the audit columns.

**Review by lucyb — APPROVED** [review]

**Pull request merged**